apps/dcos_dns/src/dcos_dns_mesos.erl

```diff
@@ -0,0 +1,384 @@
+-module(dcos_dns_mesos).
+-behavior(gen_server).
+
+-include("dcos_dns.hrl").
+-include_lib("dns/include/dns.hrl").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%% API
+-export([
+    start_link/0,
+    dns_records/2,
+    push_zone/2
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2,
+    handle_info/2, terminate/2, code_change/3]).
+
+-type task() :: dcos_net_mesos_state:task().
+-type task_id() :: dcos_net_mesos_state:task_id().
+
+-define(DCOS_DNS_TTL, 5).
+-define(IS_RUNNING(TS), not is_boolean(TS)).
+
+-record(state, {
+    ref :: reference(),
+    tasks :: #{ task_id() => [dns:dns_rr()] },
+    masters_ref :: reference(),
+    masters = [] :: [dns:dns_rr()],
+    ops_ref = make_ref() :: reference(),
+    ops = [] :: [riak_dt_orswot:orswot_op()]
+}).
+-type state() :: #state{}.
+
+-spec(start_link() -> {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    self() ! init,
+    {ok, []}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info(init, State) ->
+    State0 = handle_init(State),
+    {noreply, State0};
+handle_info({task_updated, Ref, TaskId, Task}, #state{ref=Ref}=State) ->
+    ok = dcos_net_mesos_state:next(Ref),
```

Comment:
what is this for?

Reply:
it's a part of back-pressure logic. See https://github.com/dcos/dcos-net/pull/37#discussion_r181465381